From: Heinz Mauelshagen <heinzm@redhat.com>
Date: Tue, 23 Feb 2010 11:55:44 -0500
Subject: [dm] raid45 target: constructor error path oops fix
Message-id: <1266926144.21349.23.camel@o>
Patchwork-id: 23405
O-Subject: [RHEL6.0 PATCH] dm: raid45 target: constructor error path oops fix
Bugzilla: 567605
RH-Acked-by: Bryn M. Reeves <bmr@redhat.com>
RH-Acked-by: Jonathan E Brassow <jbrassow@redhat.com>

RHEL6.0 device mapper: raid45 target

The RAID target oopses in el6 kernels on mapping table loads in the
error path, because of an unconditional release of a failed dm_io_client
creation in stripe_recover_free().

Resolves: bz#567605 (fix ported to el6 as seen in el5.5 bz#565494)

Please ACK.

Signed-off-by: Heinz Mauelshagen <heinzm@redhat.com>
Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/drivers/md/dm-raid45.c b/drivers/md/dm-raid45.c
index 0ad87e9..80e068a 100644
--- a/drivers/md/dm-raid45.c
+++ b/drivers/md/dm-raid45.c
@@ -1455,7 +1455,7 @@ static void stripe_recover_free(struct raid_set *rs)
 
 	mc = rec->mem_cache_client;
 	rec->mem_cache_client = NULL;
-	if (mc) {
+	if (mc && !IS_ERR(mc)) {
 		struct stripe *stripe;
 
 		while (!list_empty(&rec->stripes)) {
@@ -1467,8 +1467,11 @@ static void stripe_recover_free(struct raid_set *rs)
 		}
 
 		dm_mem_cache_client_destroy(mc);
-		dm_io_client_destroy(rec->dm_io_client);
-		rec->dm_io_client = NULL;
+
+		if (rec->dm_io_client && !IS_ERR(rec->dm_io_client)) {
+			dm_io_client_destroy(rec->dm_io_client);
+			rec->dm_io_client = NULL;
+		}
 	}
 }
 
